### PR TITLE
Refactor build steps runners

### PR DESCRIPTION
**Pull request by TadCordle**
Part of #262

### CODE REVIEW

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/BuildDockerSteps.java [hunk 1]
```diff
@@ -54,12 +56,35 @@ public BuildDockerSteps(
     this.buildConfiguration = buildConfiguration;
     this.sourceFilesConfiguration = sourceFilesConfiguration;
     this.cachesInitializer = cachesInitializer;
+    startupMessage =
```
Comment (coollog) [inline jib-core/src/main/java/com/google/cloud/tools/jib/builder/BuildDockerSteps.java]: Can make this a `private static final` format and call `String.format` with the target image reference.

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/BuildSteps.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ */
+
+package com.google.cloud.tools.jib.builder;
+
+import com.google.cloud.tools.jib.cache.CacheDirectoryNotOwnedException;
+import com.google.cloud.tools.jib.cache.CacheMetadataCorruptedException;
+import java.io.IOException;
+import java.util.concurrent.ExecutionException;
+
+/** Steps run for building an image. */
```
Comment (coollog) [inline jib-core/src/main/java/com/google/cloud/tools/jib/builder/BuildSteps.java]: `Steps for...`?

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/BuildSteps.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+public interface BuildSteps {
+  void run()
```
Comment (coollog) [inline jib-core/src/main/java/com/google/cloud/tools/jib/builder/BuildSteps.java]: newline

File: jib-core/src/main/java/com/google/cloud/tools/jib/frontend/BuildStepsRunner.java [hunk 4]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2018 Google LLC. All rights reserved.
+ * Copyright 2018 Google LLC. All Rights Reserved.
```
Comment (coollog) [inline jib-core/src/main/java/com/google/cloud/tools/jib/frontend/BuildStepsRunner.java]: Looks like some capitalization happened here.

File: jib-core/src/main/java/com/google/cloud/tools/jib/frontend/BuildStepsRunner.java [hunk 5]
```diff
@@ -37,21 +39,49 @@
...
+   */
+  public static BuildStepsRunner forBuildImage(
+      BuildConfiguration buildConfiguration,
+      SourceFilesConfiguration sourceFilesConfiguration,
+      Path cacheDirectory,
+      boolean useOnlyProjectCache)
+      throws CacheDirectoryCreationException {
+    Initializer cacheInitializer = getCacheInitializer(cacheDirectory, useOnlyProjectCache);
+    return new BuildStepsRunner(
+        () -> new BuildImageSteps(buildConfiguration, sourceFilesConfiguration, cacheInitializer));
```
Comment (coollog) [inline jib-core/src/main/java/com/google/cloud/tools/jib/frontend/BuildStepsRunner.java]: If this is just passing in a new `BuildImageSteps`, it doesn't seem like we need the supplier.

File: jib-core/src/main/java/com/google/cloud/tools/jib/frontend/BuildStepsRunner.java [hunk 6]
```diff
@@ -64,79 +94,74 @@ public static BuildImageStepsRunner newRunner(
...
+    return cachesInitializer;
   }
 
-  private final Supplier<BuildImageSteps> buildImageStepsSupplier;
-
-  @VisibleForTesting
-  BuildImageStepsRunner(Supplier<BuildImageSteps> buildImageStepsSupplier) {
-    this.buildImageStepsSupplier = buildImageStepsSupplier;
+  /** @return true if Docker is installed on the user's system. */
+  public static boolean isDockerInstalled() {
```
Comment (coollog) [inline jib-core/src/main/java/com/google/cloud/tools/jib/frontend/BuildStepsRunner.java]: Move above private static method.

File: jib-core/src/main/java/com/google/cloud/tools/jib/frontend/BuildStepsRunner.java [hunk 7]
```diff
@@ -178,7 +203,7 @@ public void buildImage(HelpfulSuggestions helpfulSuggestions)
     }
```
Comment (coollog) [inline jib-core/src/main/java/com/google/cloud/tools/jib/frontend/BuildStepsRunner.java]: We prob want to change `BuildImageStepsExecutionException` to `BuildStepsExecutionException`.

File: jib-core/src/main/java/com/google/cloud/tools/jib/frontend/BuildStepsRunner.java [hunk 8]
```diff
@@ -178,7 +203,7 @@ public void buildImage(HelpfulSuggestions helpfulSuggestions)
...
   }
 
-  private void handleRegistryUnauthorizedException(
+  private static void handleRegistryUnauthorizedException(
```
Comment (coollog) [inline jib-core/src/main/java/com/google/cloud/tools/jib/frontend/BuildStepsRunner.java]: static methods above nonstatic

File: jib-core/src/main/java/com/google/cloud/tools/jib/frontend/BuildStepsRunner.java (hunk 4 above)
Comment (TadCordle) [reply jib-core/src/main/java/com/google/cloud/tools/jib/frontend/BuildStepsRunner.java]: It looks like this happens in a few places in the project, not just this PR. I'll fix the rest in a follow-up.

File: jib-core/src/main/java/com/google/cloud/tools/jib/frontend/BuildStepsRunner.java (hunk 6 above)
Comment (TadCordle) [reply jib-core/src/main/java/com/google/cloud/tools/jib/frontend/BuildStepsRunner.java]: Is there a style guide somewhere for the ordering? I keep forgetting it.

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/BuildDockerSteps.java [hunk 9]
```diff
@@ -42,6 +42,9 @@
 public class BuildDockerSteps implements BuildSteps {
 
   private static final String DESCRIPTION = "Building and pushing image";
+  private static final String STARTUP_FORMAT = "Building to Docker daemon as %s...";
```
Comment (coollog) [inline jib-core/src/main/java/com/google/cloud/tools/jib/builder/BuildDockerSteps.java]: nit: `STARTUP_MESSAGE_FORMAT`?

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/BuildDockerSteps.java [hunk 10]
```diff
@@ -42,6 +42,9 @@
...
+  private static final String SUCCESS_FORMAT =
```
Comment (coollog) [inline jib-core/src/main/java/com/google/cloud/tools/jib/builder/BuildDockerSteps.java]: A comment about this being cyan.

File: jib-core/src/main/java/com/google/cloud/tools/jib/frontend/BuildStepsRunner.java [hunk 11]
```diff
@@ -36,13 +36,12 @@
...
 import java.nio.file.Path;
 import java.util.concurrent.ExecutionException;
-import java.util.function.Supplier;
 import org.apache.http.conn.HttpHostConnectException;
 
 /** Runs a {@link BuildSteps} and builds helpful error messages. */
 public class BuildStepsRunner {
 
-  private final Supplier<BuildSteps> buildStepsSupplier;
+  private final BuildSteps buildSteps;
```
Comment (coollog) [inline jib-core/src/main/java/com/google/cloud/tools/jib/frontend/BuildStepsRunner.java]: Instance fields should go below statics

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/BuildDockerSteps.java [hunk 12]
```diff
@@ -42,8 +42,9 @@
 public class BuildDockerSteps implements BuildSteps {
 
   private static final String DESCRIPTION = "Building and pushing image";
-  private static final String STARTUP_FORMAT = "Building to Docker daemon as %s...";
-  private static final String SUCCESS_FORMAT =
+  private static final String STARTUP_MESSAGE_FORMAT = "Building to Docker daemon as %s...";
+  // String parameter in cyan.
```
Comment (coollog) [inline jib-core/src/main/java/com/google/cloud/tools/jib/builder/BuildDockerSteps.java]: Should prob be explicit that the format is used with a target image reference.

**Pull request merged**